File: packages/oath/ios/OathErrorCodes.swift
/*
 * Copyright (c) 2026 Ping Identity Corporation. All rights reserved.
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file for details.
 */

import Foundation

/// Stable error codes emitted by the OATH module.
///
/// Keep these in sync with JS `OathErrorCodes` and Android
/// `OathErrorCodes`. Each case maps to a unique string that JS
/// consumers can match on to provide user-facing error handling.
internal enum OathErrorCodes: String {
  /// The requested credential was not found in storage.
  case credentialNotFound = "OATH_CREDENTIAL_NOT_FOUND"
  /// The credential is locked and cannot be used without authentication.
  case credentialLocked = "OATH_CREDENTIAL_LOCKED"
  /// A credential with the same identifier already exists.
  case duplicateCredential = "OATH_DUPLICATE_CREDENTIAL"
  /// The provided URI is not a valid OATH URI.
  case invalidUri = "OATH_INVALID_URI"
  /// A required parameter is missing from the request.
  case missingParameter = "OATH_MISSING_PARAMETER"
  /// A provided parameter has an invalid value.
  case invalidParameter = "OATH_INVALID_PARAMETER"
  /// The URI could not be formatted or parsed correctly.
  case uriFormatting = "OATH_URI_FORMATTING"
  /// OATH code generation failed for the credential.
  case codeGenerationFailed = "OATH_CODE_GENERATION_FAILED"
  /// The operation violates an OATH policy constraint.
  case policyViolation = "OATH_POLICY_VIOLATION"
  /// The OATH client failed to initialize.
  case initializationFailed = "OATH_INITIALIZATION_FAILED"
  /// The cleanup operation failed.
  case cleanupFailed = "OATH_CLEANUP_FAILED"
  /// A storage operation failed.
  case storageFailed = "OATH_STORAGE_FAILURE"
  /// The credential storage is corrupted.
  case storageCorrupted = "OATH_STORAGE_CORRUPTED"
  /// Access to the credential storage was denied.
  case storageAccessDenied = "OATH_STORAGE_ACCESS_DENIED"
  /// An unexpected state was encountered.
  case stateError = "OATH_STATE_ERROR"
  /// A generic, unclassified error.
  case unknown = "OATH_UNKNOWN_ERROR"
}

File: packages/oath/android/src/main/java/com/pingidentity/rnoath/OathErrorCodes.kt
/*
 * Copyright (c) 2026 Ping Identity Corporation. All rights reserved.
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file for details.
 */
package com.pingidentity.rnoath

/**
 * Stable error codes emitted by the OATH module.
 *
 * @remarks
 * Keep these in sync with JS `OathErrorCode` and iOS `OathErrorCodes`.
 *
 * Four codes present on iOS are intentionally absent here:
 * `OATH_MISSING_PARAMETER`, `OATH_CLEANUP_FAILED`, `OATH_STORAGE_CORRUPTED`,
 * and `OATH_STORAGE_ACCESS_DENIED`. The Android MFA commons library has no
 * equivalent exception types for these conditions, so Android cannot emit them.
 * The TypeScript `OathErrorCode` union includes all codes from both platforms
 * by design.
 */
object OathErrorCodes {
  const val OATH_INVALID_URI = "OATH_INVALID_URI"
  const val OATH_INVALID_PARAMETER = "OATH_INVALID_PARAMETER"
  const val OATH_CREDENTIAL_NOT_FOUND = "OATH_CREDENTIAL_NOT_FOUND"
  const val OATH_CREDENTIAL_LOCKED = "OATH_CREDENTIAL_LOCKED"
  const val OATH_DUPLICATE_CREDENTIAL = "OATH_DUPLICATE_CREDENTIAL"
  const val OATH_CODE_GENERATION_FAILED = "OATH_CODE_GENERATION_FAILED"
  const val OATH_INITIALIZATION_FAILED = "OATH_INITIALIZATION_FAILED"
  const val OATH_STORAGE_FAILURE = "OATH_STORAGE_FAILURE"
  const val OATH_POLICY_VIOLATION = "OATH_POLICY_VIOLATION"
  const val OATH_STATE_ERROR = "OATH_STATE_ERROR"
  const val OATH_UNKNOWN_ERROR = "OATH_UNKNOWN_ERROR"
}
